2단계 - 문자열 덧셈 계산기

ghtjr410 opened the pull request:
## 문자열 덧셈 계산기 과제 제출

안녕하세요!
문자열 덧셈 계산기 과제 제출합니다.

### 구현 과정

슬랙에서 가능하면 혼자 힘으로 구현하라는 말씀을 듣고, 최대한 **TDD 사이클**을 따라 진행했습니다.

1. **Red**: 요구사항이 늘어날 때마다 테스트 케이스 추가
2. **Green**: 해당 테스트만 통과하는 최소한의 코드 작성
3. **Refactor**: SRP 원칙을 고민하며 메서드 분리

### 고민했던 부분

- **점진적 구현**: 처음부터 완벽한 구조를 만들려 하지 않고, 테스트를 통과시키며 자연스럽게 구조 개선
- **SRP 준수**: 각 메서드가 단일 책임만 가지도록 지속적으로 리팩터링
- **메서드 네이밍**: 의도를 명확히 드러내는 이름 짓기에 많은 고민

### 최종 구조

- 고수준 메서드(`splitAndSum`)는 전체 흐름만 표현
- 세부 구현은 각각의 작은 메서드들로 분리
- Stream API를 활용한 함수형 스타일 적용

리뷰 잘 부탁드립니다!

ghtjr410 committed:
refactor: 피드백 반영 - 불필요한 로컬 변수 제거 및 테스트 코드 간결화

StringTest.java
- 재사용되지 않는 로컬 변수 `input`, `targetIndex`, `outOfBoundIndex` 제거
- 문자열 리터럴을 직접 사용하여 inline 처리
- 예외 테스트에서 인덱스 값 및 메시지 검증 수정

ghtjr410 committed:
feat: 빈 문자열/null 처리 - 0 반환

StringCalculator.java
- 문자열 덧셈 계산기 클래스 생성
- 입력값이 null이거나 빈 문자열일 경우 0을 반환하도록 구현

StringCalculatorTest.java
- 빈 문자열 입력 시 0을 반환하는 테스트 케이스 추가

ghtjr410 committed:
feat: 숫자 하나 입력 시 해당 숫자 반환

StringCalculator.java
- 숫자 문자열을 정수로 변환하여 반환하도록 구현

StringCalculatorTest.java
- 숫자 하나 입력 시 해당 숫자를 반환하는 테스트 케이스 추가

ghtjr410 committed:
feat: 쉼표 구분자로 여러 숫자 합산 기능 추가

StringCalculator.java
- 쉼표로 구분된 숫자들을 split하여 합산하도록 구현

StringCalculatorTest.java
- 쉼표 구분자 테스트 케이스 추가

ghtjr410 committed:
feat: 콜론 구분자 지원 추가

StringCalculator.java
- 쉼표와 콜론 모두 구분자로 인식하도록 정규식 패턴 수정

StringCalculatorTest.java
- 콜론 구분자 테스트 케이스 추가

ghtjr410 committed:
feat: 커스텀 구분자 지원 추가

StringCalculator.java
- "//{구분자}\n{숫자}" 형식의 커스텀 구분자 파싱 기능 구현
- 커스텀 구분자를 기존 패턴에 추가하여 split 처리

StringCalculatorTest.java
- 커스텀 구분자 테스트 케이스 추가

ghtjr410 committed:
feat: 음수 입력 시 예외 처리 추가

StringCalculator.java
- 음수 입력 시 RuntimeException 발생하도록 검증 로직 추가
- 예외 메시지에 입력된 음수 값 포함

StringCalculatorTest.java
- 음수 입력 시 예외 발생 테스트 케이스 추가

ghtjr410 committed:
refactor: 변수명 가독성 개선

StringCalculator.java
- value → input (입력값 의미 명확화)
- pattern → delimiterPattern (구분자 패턴임을 명시)
- delimiterEndIndex → newLineIndex (개행 문자 위치 의미 명확화)
- result → tokens (split 결과가 토큰 배열임을 명시)
- s → token (단일 토큰 의미 명확화)

ghtjr410 committed:
refactor: null/빈 문자열 검증 로직 메서드 분리

StringCalculator.java
- isNullOrEmpty 메서드 추가로 검증 로직 명확화

ghtjr410 committed:
refactor: 커스텀 구분자 존재 여부 검증 로직 메서드 분리

StringCalculator.java
- hasCustomDelimiter 메서드 추가로 검증 의도 명확화

ghtjr410 committed:
refactor: 커스텀 구분자 파싱 로직 메서드 분리

StringCalculator.java
- extractCustomDelimiter: 커스텀 구분자 추출 로직 분리
- extractNumbersPart: 실제 숫자 문자열 추출 로직 분리
- getDelimiterEndPosition: 개행 문자 위치 탐색 로직 분리

ghtjr410 committed:
refactor: 숫자 파싱 및 합산 로직을 Stream으로 개선

StringCalculator.java
- parsePositiveNumber: 문자열을 양수로 변환하는 로직 분리
- Stream API 활용으로 숫자 변환과 합산 로직 간결화

ghtjr410 committed:
refactor: SRP 원칙에 따라 메서드 책임 분리 및 상수 추출

StringCalculator.java
- 상수 추출: DEFAULT_DELIMITER_PATTERN, CUSTOM_DELIMITER_PREFIX
- determineDelimiterPattern: 구분자 패턴 결정 로직 분리
- addCustomDelimiter: 커스텀 구분자 추가 로직 분리
- extractNumbersPart: 숫자 문자열 추출 로직 개선 (기본 구분자 처리 포함)
- sumTokens: 토큰 배열 합산 로직 분리
- validatePositive: 양수 검증 로직 분리
- isNegative: 음수 판별 로직 분리
- splitAndSum: 고수준 흐름만 표현하도록 단순화

javajigi commented on src/main/java/caculator/StringCalculator.java:
```diff
@@ -0,0 +1,79 @@
+    private static final String DEFAULT_DELIMITER_PATTERN = ",|:";
+    private static final String CUSTOM_DELIMITER_PREFIX = "//";
+
+    public static int splitAndSum(String input) {
```
👍

javajigi commented on src/main/java/caculator/StringCalculator.java:
```diff
@@ -0,0 +1,79 @@
+    private static final String CUSTOM_DELIMITER_PREFIX = "//";
+
+    public static int splitAndSum(String input) {
+        if (isNullOrEmpty(input)) {
```
if body가 한 라임에도 극단적으로 분리한 것이 인상적임 👍

javajigi commented on src/main/java/caculator/StringCalculator.java:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        String delimiterPattern = determineDelimiterPattern(input);
+        String numbersPart = extractNumbersPart(input);
```
현재 구현도 넘 좋음
단, 이 2개의 메서드는 둘이 연결된 느낌도 든다.
예를 들어 추상화 수준에서 보면 `문자열을 던져줄테니 split()해. 나는 어떻게 split하는지는 관심없으니 니가 알아서 처리해.`와 같은 느낌이 들도록 메서드를 리팩터링해 보는 것은 어떨까?

ghtjr410 replied:
피드백 감사합니다!
추상화 수준 통일에 대해 고민하며 진행해보겠습니다!

javajigi commented on src/main/java/caculator/StringCalculator.java:
```diff
@@ -0,0 +1,79 @@
+        return input.indexOf("\n");
+    }
+
+    private static int sumTokens(String[] tokens) {
```
이 메서드는 stream을 사용하기 때문에 상당히 간결한 느낌을 주고 있음.
단, 이 메서드는 한 가지 책임만 가지고 있는 것이 맞는가?
그럼 다름 측면으로 고민해 봤을 때 int[]에 대한 sum을 구하고 싶을 때 재사용 가능한 구조를 만들려면 어떻게 하면 좋을까?

ghtjr410 replied:
피드백 감사합니다! 
재사용성과 단일 책임에 대해 고민하며 진행해보겠습니다!

javajigi reviewed (changes requested):
2단계의 의도에 맞춰 메서드를 정말 극단적으로 리팩터링했네요. 👍 
시작 단계에서 tdd 사이클에 맞춰 개발하고 리팩터링하기 쉽지 않았을텐데 tdd 사이클을 지키면서 개발했다는 것이 더 인상적이네요.
지금 단계도 충분히 많은 연습을 한 것으로 보여 바로 merge하려다 여기서 한 단계 더 나아가보는 연습을 해봤으면 하는 바람으로 피드백 남겨 봅니다.

ghtjr410 committed:
refactor: 추상화 수준 통일을 위해 parseToTokens 메서드 추가

StringCalculator.java
- parseToTokens: 구분자 처리 및 토큰 분리 로직을 하나의 메서드로 캡슐화
- buildCustomDelimiterPattern: determineDelimiterPattern에서 이름 변경 (역할 명확화)
- splitAndSum: 고수준 추상화 수준으로 단순화 (어떻게 파싱하는지 숨김)
- extractNumbersPart: 불필요한 조건문 제거 (parseToTokens에서 이미 처리)

ghtjr410 committed:
refactor: 변환과 합산 로직 분리로 재사용성 개선

StringCalculator.java
- toPositiveNumbers: 문자열 배열을 숫자 배열로 변환하는 로직 분리
- sum: 숫자 배열 합산 로직 분리 (int[] 배열에 대해 재사용 가능)
- sumTokens: 변환과 합산을 명확히 구분하여 단일 책임 원칙 준수

ghtjr410 commented:
피드백 감사합니다!

두 가지 피드백 모두 반영했습니다.

### 1. 추상화 수준 통일
- `parseToTokens()` 메서드로 파싱 로직 완전히 캡슐화
- `splitAndSum()`에서는 "어떻게" 파싱하는지 알 필요 없도록 개선

### 2. 재사용 가능한 구조
- `toPositiveNumbers()`: 문자열 배열 → 숫자 배열 변환
- `sum()`: 숫자 배열 합산 (독립적으로 재사용 가능!)
- 변환과 합산 로직 완전 분리

다시 리뷰 부탁드립니다!

javajigi commented on src/main/java/caculator/StringCalculator.java:
```diff
@@ -12,30 +12,33 @@ public static int splitAndSum(String input) {
-        String[] tokens = numbersPart.split(delimiterPattern);
-
+        String[] tokens = parseToTokens(input);
         return sumTokens(tokens);
```
정말 빠르게 피드백 반영 잘 했네요. 👍 
이미 메서드 분리는 너무 잘한 상태라서요.
메서드를 다음과 같이 연결하면 어떨까요?

- token을 메서드 이름으로 계속 사용하는 것도 굳이 불필요할 것 같아 제거
- positive도 숫자의 의미를 포함하고 있어 numbers 제거
- 굳이 로컬 변수가 필요 없을 것 같아 메서드 체이닝으로 구현

```
return sum(toPositive(parse(input)));
```

ghtjr410 replied:
피드백 감사합니다!

말씀해주신 대로 다시 보니 `parse → toPositive → sum`의 흐름이 
선형적인 데이터 변환 파이프라인이라고 생각이 듭니다.

각 단계의 출력이 다음 단계의 입력이 되는 구조를
메서드 체이닝으로 표현하니 흐름이 훨씬 명확해졌습니다.

정말 많이 배웠습니다. 감사합니다!

javajigi commented on src/main/java/caculator/StringCalculator.java:
```diff
@@ -56,9 +56,14 @@ private static int getDelimiterEndPosition(String input) {
+                .toArray();
     }
 
     private static int parsePositiveNumber(String token) {
```
메서드 분리를 넘 잘하고 있고, 다른 분들에 비해 속도가 빨라 클래스 분리 도전한번 해보면 어떨까?
`규칙 3: 모든 원시값과 문자열을 포장한다.` 원칙을 지켜 리팩터링 한번 도전해 보면 어떨까?
즉, token 문자열을 포장하고, 포장한 객체가 로직을 수행하도록 리팩터링해본다.
이때의 결과는 현재와 어떻게 다를까?

ghtjr410 replied:
피드백 감사합니다!
객체지향 생활 체조 원칙의 원시값 포장 규칙 도전해보겠습니다! 
정말 많이 배우고 있습니다. 감사합니다!

javajigi reviewed (changes requested):
피드백의 의도를 정말 잘 파악해 피드백 반영 넘넘 잘 했네요. 👍 
메서드 분리 관점에서 피드백 하나 남겼고요.

메서드 분리는 이미 일정 수준 잘하고 있고, 다른 분들에 비해 속도가 빨라 이 작은 미션에서 클래스 분리 도전을 위한 피드백 한번 남겨봤어요.
한번 도전해 보면 어떨까요?

ghtjr410 committed:
refactor: 메서드 체이닝으로 가독성 개선

StringCalculator.java
- parseToTokens → parse: 메서드명 간결화
- toPositiveNumbers → toPositive: 메서드명 간결화
- sum(toPositive(parse(input))): 메서드 체이닝으로 흐름 명확화
- 중간 변수 제거로 코드 간결화

ghtjr410 committed:
refactor: 문자열을 Token 객체로 포장하여 검증 책임 이전

StringCalculator.java
- 기존 parsePositiveNumber, validatePositive, isNegative 메서드 제거
- Token 객체를 활용해 문자열 검증 및 양수 변환 수행

Token.java
- 문자열을 포장하는 Value Object 도입
- null, 공백, 비숫자 입력 검증 로직 추가 (validate 메서드 분리)
- toPositive: 음수 검증 및 예외 처리 책임 담당
- of(String) 팩토리 메서드 추가로 생성 시 명확한 의도 표현

TokenTest.java
- 정상/비정상 입력값에 대한 단위 테스트 추가
  - null, 공백, 비숫자 입력 시 예외 검증
  - 양수/0/음수 변환 시 동작 검증
- @ParameterizedTest로 공백 입력 케이스 가독성 향상

ghtjr410 committed:
refactor: Token 배열 변환 책임을 Tokens 객체로 위임

StringCalculator.java
- Token 변환 및 양수 검증 로직을 Tokens.from(tokens).toPositive()로 대체
- StringCalculator의 책임을 단순 합산 조립 역할로 축소

Tokens.java
- 신규 Tokens 도메인 추가
  - Token 리스트를 감싸는 불변 VO로 설계
  - null, empty 검증 로직 추가
  - String 배열 → Token 리스트 변환 정적 팩토리 메서드(from) 구현
  - Token 리스트 → int 배열 변환 메서드(toPositive) 구현

TokensTest.java
- Tokens 생성 및 변환 로직 단위 테스트 추가
  - 정상 케이스: 배열 입력 시 Token 리스트 생성 검증
  - 예외 케이스: null, 빈 배열, 음수 포함 시 예외 발생 검증
  - 양수 변환 및 예외 메시지 일관성 검증

ghtjr410 committed:
feat: Numbers 일급 컬렉션 추가로 숫자 배열 합산 책임 분리

Numbers.java
- 숫자 컬렉션을 감싸는 일급 컬렉션
- 불변성 보장 (List.copyOf)
- null 및 빈 리스트 검증 (Fail-Fast)
- sum() 메서드로 합산 책임 캡슐화

NumbersTest.java
- Numbers 생성 및 합산 테스트
- null, 빈 리스트, 불변성 검증

StringCalculator.java
- Numbers 객체에 합산 책임 위임
- 원시값 포장 원칙 적용

ghtjr410 committed:
refactor: Tokens 불변성 보장 및 테스트 개선

Tokens.java
- List.copyOf()로 불변성 보장 추가
- from() 메서드 예외 메시지 명확화 (토큰 리스트 → 토큰 배열)

TokensTest.java
- 불변성 보장 테스트 추가
- 테스트 순서 재정렬 (생성자 → from → 기능)

ghtjr410 committed:
test: Token 생성자 테스트 추가

TokenTest.java
- 생성자를 통한 Token 생성 테스트 추가
- null, 공백, 숫자 형식 검증 테스트 추가
- 테스트 순서 재정렬 (생성자 → of → toPositive)

ghtjr410 committed:
refactor: 예외 메시지 일관성 개선 및 테스트 보완

Token.java, Tokens.java, Numbers.java
- 예외 메시지를 "XX는 YY일 수 없습니다" 패턴으로 통일
- "리스트" → "목록"으로 용어 변경 (도메인 언어)
- "숫자가 아닙니다" → "숫자 형식이 아닙니다"로 명확화
- "음수를 입력할" → "음수는 입력할"로 일관성 개선

TokenTest.java, TokensTest.java, NumbersTest.java
- 예외 메시지 검증 업데이트
- Numbers.from() 빈 배열 테스트 추가
- 테스트 메서드명 간결화 (from_빈_문자열_배열이면_IllegalArgumentException_발생 → from_빈_배열)

ghtjr410 commented:
피드백 반영했습니다!

### 구현 과정에서의 고민

이번 리팩터링을 진행하며 **객체의 책임과 경계**에 대해 깊이 고민했습니다.

#### 1. 책임의 경계 설정
각 객체가 "무엇을 알아야 하는가"와 "무엇을 할 수 있어야 하는가"를 명확히 구분하려 노력했습니다.
- `Token`: 자신이 유효한 숫자 문자열인지 검증하고, 양수로 변환하는 책임
- `Tokens`: Token 컬렉션을 관리하고, 전체를 양수 배열로 변환하는 책임
- `Numbers`: 숫자 컬렉션을 관리하고, 합산하는 책임

#### 2. 불변식 재정립
객체는 외부 세계를 알 수 없기 때문에, 각 객체가 생성되는 시점에 자신의 불변식을 스스로 검증해야 한다고 생각했습니다.

예를 들어 `Tokens.from(String[] tokens)`에서:
```java
public static Tokens from(String[] tokens) {
    Objects.requireNonNull(tokens, "토큰 배열은 null일 수 없습니다.");
    // tokens가 null이 아님을 보장한 상태로 생성자에 전달
}
```

외부에서 받은 배열이 null이 아님을 명시적으로 검증한 후, 내부 생성자로 전달합니다.
이는 "이 시점부터는 null이 아닌 배열임을 보장한다"는 계약을 명확히 하기 위함입니다.

#### 3. 정적 팩토리 메서드 선호
개인적으로 `new` 연산자보다 정적 팩토리 메서드를 선호합니다.

**이유:**
- **의미 전달**: `Token.of("1")`이 `new Token("1")`보다 "값으로부터 생성"이라는 의도가 명확
- **타입 전환 시점의 검증**: `Tokens.from(String[])`, `Numbers.from(int[])`처럼 다른 타입으로부터 객체를 생성할 때, 변환 과정에서 필요한 검증을 명시적으로 수행 가능
- **계약 명확화**: 배열이 null인지 검증하는 책임을 정적 팩토리 메서드가 가짐으로써, "이 메서드를 거치면 유효한 객체"라는 계약을 코드로 표현

#### 4. TDD와 객체 설계의 긴장

처음에는 "기능을 구현"하는 관점으로 TDD를 접근하다 보니 객체 설계가 제대로 되지 않았습니다.
- 테스트: "1,2를 입력하면 3이 나온다" 
- 구현: StringCalculator에 모든 로직을 때려박기

하지만 "**객체의 책임과 행동**"에 집중하여 접근하니 점차 나은 방향으로 나아가는 것을 느꼈습니다.
- 테스트: "Token은 자신을 양수로 변환할 수 있어야 한다"
- 구현: Token 객체가 toPositive() 책임을 가짐

### 질문

객체지향 관점에서 TDD를 진행할 때, "기능 테스트"가 아닌 "**객체의 책임과 행동 테스트**"로 사고를 전환하는 것이 핵심이라고 느꼈습니다.

이러한 사고 전환이 좋은 접근 방식인지, 괜찮은 접근 방식이라면 더 자연스럽게 할 수 있는 방법이나 팁이 있을까요?

정말 많이 배웠습니다. 다시 리뷰 부탁드립니다!

javajigi commented on src/main/java/caculator/Numbers.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.List;
+import java.util.Objects;
+
+public record Numbers(List<Integer> values) {
```
일급 콜렉션 👍

javajigi commented on src/main/java/caculator/Numbers.java:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    public static Numbers from(int[] values) {
```
굳이 정적 팩터리 메서드로 구현할 필요가 있을까?
이번 기회에 생성자와 정적 팩터리 메서드의 적절한 사용에 대해 고민해 보는 기회를 만들면 어떨까?

AI에게 정적 팩터리 메서드 단점에 대해 질문한 결과
1. 🔍 발견성 (Discoverability) 문제
생성자와의 차이점: 생성자는 자바 API 문서(Javadoc)에서 클래스 이름 아래에 명확하게 표시됩니다. 개발자는 new 키워드를 사용하는 순간 IDE의 자동 완성 기능을 통해 사용 가능한 모든 생성자를 쉽게 확인할 수 있습니다.

정적 팩터리 메서드의 문제: 정적 팩터리 메서드는 일반적인 static 메서드와 동일하게 취급되므로, 개발자가 API 문서나 코드를 직접 살펴보지 않으면 해당 클래스가 객체 생성을 위한 특별한 메서드(팩터리 메서드)를 제공하는지 쉽게 알 수 없습니다.

완화 방법: 이펙티브 자바(Effective Java)에서 제시하는 from, of, valueOf, getInstance 등의 표준 명명 관례를 따르면 이 문제가 완화될 수 있지만, 여전히 개발자가 이 관례를 알고 있어야 합니다.

2. 🚫 상속 제한
필요성: 클래스를 상속(확장)하려면 하위 클래스의 생성자가 상위 클래스의 public 또는 protected 생성자에 접근할 수 있어야 합니다.

정적 팩터리의 영향: 정적 팩터리 메서드를 사용하는 클래스는 객체 생성 로직을 캡슐화하고 재사용 가능한 인스턴스를 관리하기 위해 생성자를 private으로 선언하는 경우가 일반적입니다.

결과: 생성자가 private이면 외부에서 상속이 불가능해집니다. 즉, 정적 팩터리 메서드만 제공하는 클래스는 상속을 통한 확장이 불가능합니다.

참고: 이는 불변(Immutable) 객체를 설계하거나 상속보다 조합(Composition)을 권장하는 설계 원칙에서는 오히려 장점으로 간주되기도 합니다.

3. 🧩 인터페이스 타입으로 구현할 수 없음
자바의 인터페이스는 정적 메서드를 가질 수 있지만 (Java 8 이후), 인터페이스의 인스턴스를 생성하는 정적 팩터리 메서드를 정의하는 것은 불가능합니다.

예외: 자바는 관례적으로 Collections.singletonList()처럼 **별도의 동반 클래스(companion class)**를 만들어 인터페이스의 인스턴스를 반환하는 정적 팩터리 메서드를 제공합니다. (예: java.util.Collections 클래스는 java.util.Collection 인터페이스를 위한 정적 팩터리 메서드들을 모아둔 클래스입니다.)

ghtjr410 replied:
피드백 감사합니다!
정적 팩터리 메서드의 단점과 생성자 선택 기준을 다시 고민해볼 수 있는 좋은 기회였습니다.

### 정적 팩터리를 선택한 기준
> 핵심은 "변환 책임을 어디에 둘 것인가?"였습니다.
``` java
// 필드 타입과 외부 타입이 다를 때 -> from
public static Tokens from(String[] tokens) // String[] -> List<Token>
public static Numbers from(int[] values) // int[] -> List<Integer>

// 동일할 때 -> of
public static Token of(String value) // String -> String (검증만)
```
### `from`: 변환 책임의 캡슐화
외부 타입을 내부 구조로 변환하는 책임을 객체 스스로 가져야 한다고 판단했습니다.
``` java
// 변환 책임이 객체 내부로 캡술화됨
Tokens tokens = Tokens.from(stringArray);

// vs. 변환 책임이 호출부에 분산
List <Token> tokenList = Arrays.stream(stringArray)
                                                    .map(Token::new)
                                                    .toList();
Tokens tokens = new Tokens(tokenList);
```
- 의사표현의 명확성: "문자열 배열로부터 Tokens를 생성한다"는 의도를 명확히 표현
- 캡슐화 강화: 호출부는 변환 과정을 몰라도 됨
### `of`: 생성자와 동일한 역할
#### `Token.of(value)`는 사실상 `new Token(value)`와 기능적으로 동일합니다.
``` java
public static Token of(String value) {
        return new Token(value); // 단순 생성자 위임
}
```
- 사용 이유는 `from()`과의 네이밍 일관성을 유지하기 위함이었습니다.
- 하지만 기능적 이점이 없다는 점에 있어서 피드백에 전적으로 동의합니다.
### 공감한 부분
말씀하신 단점들(발견성, 상속 제한)은 명확한 트레이드 오프입니다.
- 생성자는 IDE에서 즉시 발견되지만, 정적 메서드는 문서/코드를 봐야함
- 단순한 구조에서는 생성자만으로 충분히 명확합니다.
### 결론
- 단순 생성 -> 생성자 사용
- 타입 변환이나 의미 부여가 필요한 경우 -> `from()` 사용

현재 구조는 "변환 책임을 객체 내부로 옮긴다"는 판단에 기반했으며,
이후 실습에서는 정적 팩토리의 필요성이 분명할 때만`from`을 사용하려 합니다.

혹시 위 의견에 대해 어떻게 생각하시는지 궁금합니다.
좋은 피드백 감사드립니다.

### 개인적인 소회
현재 실무에서 혼자 백엔드를 맡다 보니, 이런 피드백과 논의가 정말 값진 경험이라고 생각합니다.
작은 부분이라도 피드백 주시면 놓치지 않고 고민하며 배우겠습니다.
이런 대화들이 제 성장의 원동력이 되고 있습니다. 감사합니다!

javajigi replied:
어제 수업 중 정답을 찾지 마라는 이야기를 했었는데요.
생성자와 정적 팩터리 메서드 사용에 대한 기준도 정답처럼 고정해 버리는 것은 아닌지 한번 생각해 보면 어떨까요?
그런 측면에서 다음과 같은 코드는 어떻게 생각하나요?

```java
public class Positive {
    private final int value;

    public Positive(String value) {
        this(Integer.parseInt(value));
    }
    
    public Positive(int value) {
        if (value < 0) {
            throw new IllegalArgumentException("음수를 허용하지 않습니다.");
        }
        this.value = value;
    }
    
}
```

생성자에서 타입 변환을 하고 있는데요.
이럴 경우 정적 팩토리 메서드가 필요할까요?

다음과 같은 기준은 어떨까요?
생성자 사용의 효용이 느껴진다면 가능하면 생성자를 우선으로 구현하고, 정적 팩터리 메서드가 정말 필요하다고 느끼는 순간에만 정적 팩터리 메서드를 사용한다.

ghtjr410 replied:
피드백 감사합니다! 
말씀하신 접근이 YAGNI 원칙에 기반한다는 점 공감합니다. 
"필요할 때 도입한다"는 관점에서 보면 생성자만으로도 충분하네요.

평소 습관적으로 코드를 작성하는 것을 의식하며 피하려 노력하지만, 
이번처럼 피드백을 받고 나서야 사고가 굳어있거나 시야가 좁았다는 걸 깨닫게 됩니다.

결국 "지금 필요한가?"를 의식적으로 스스로에게 질문하며, 
동시에 고정관념에 갇혀있진 않은지 돌아보는 습관을 들이겠습니다.

말씀대로 가능한 단순하게 시작하고, 
확장 필요성이 명확해지는 시점에 리팩터링하겠습니다.

좋은 관점 나눠주셔서 감사합니다!

javajigi commented on src/main/java/caculator/Token.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Objects;
+
+public record Token(String value) {
```
Token보다 Positive라는 이름은 어떨까?
Token 관점으로 바라보니 문자를 양수로 바꾸는 관점으로 바라보게 되는 것은 아닐까?
Positive라는 이름으로 본다면 숫자의 관점으로 바라보면서 두 숫자의 합과 같은 메서드도 가지고, 수동적이기 보다 좀 더 능동적인 객체의 관점으로 바라볼 수도 있지 않을까?

ghtjr410 replied:
`Token`이라는 이름을 정한 순간, 
제 사고가 "문자열 → 숫자 변환"이라는 단계적 관점에 갇혀버렸던 것 같습니다.

말씀하신 대로 도메인 관점에서 이름을 다시 고민해보니,
현재 요구사항을 보면:

**기능 요구사항:**
- 쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하는 경우 
  구분자를 기준으로 분리한 각 숫자의 합을 반환
  (예: "" => 0, "1,2" => 3, "1,2:3" => 6)

**예외 요구사항:**
- 문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 
  RuntimeException 예외를 throw한다.

요구사항을 보니 **0은 허용**되는 것 같습니다.

그래서 `Positive`로 시작했지만, 조금 더 고민해보니
- `Positive` (수학적으로는 0 미포함)
- `PositiveOrZero` (0 이상을 명시)
- `NonNegative` (비음수)

중에서 고민이 되는데요.

네이밍에서 부정보다 긍정이 의도를 더 잘 드러낸다고 생각해서,
`PositiveOrZero`로 리팩터링해보겠습니다!

javajigi commented on src/main/java/caculator/Token.java:
```diff
@@ -0,0 +1,46 @@
+public record Token(String value) {
+
+    public Token {
+        validate(value);
```
Positive 관점으로 바라본다면 음수값이 입력되었을 때 예외를 던져 양수임을 보장하는 객체로 존재하도록 구현할 수 있지 않을까?

javajigi commented on src/test/java/caculator/PositiveOrZeroTest.java:
```diff
@@ -0,0 +1,86 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
```
👍

javajigi reviewed (changes requested):
이번 단계에서 충분히 많은 연습을 했기 때문에 바로 merge하려고 했는데요. 
피드백 한 것보다 앞으로 학습하게 될 상당히 많은 양의 코드를 구현해 일부분에 대해 피드백 남겨봤어요. 👍 
특히 다른 것보다 생성자와 정적 팩터리 메서드 사용에 대해 한번 고민해 보면 좋겠어요.
지금과 같은 요구사항에서 굳이 정적 팩터리 메서드 사용은 필요하지 않을 것 같은데 약간 남용하는 것 같아요.
이번 기회에 생성자를 사용할 시점과 정적 팩터리 메서드를 사용하는 시점에 대해 고민해 보는 기회를 만들기를 바랍니다.

ghtjr410 committed:
refactor: Token.of() 정적 팩터리 메서드 제거

- Token.of() 메서드 삭제
- Token::of → Token::new 변경
- 중복 테스트 케이스 제거 (생성자 테스트로 통합)
- 테스트 메서드명 개선 (null_빈문자열, 공백_또는_빈_문자열)

ghtjr410 committed:
test: null과 빈 문자열 테스트를 파라미터화 테스트로 개선

StringCalculatorTest.java
- @ParameterizedTest + @NullAndEmptySource로 null/빈 문자열 테스트 통합
- 예외 메시지 오타 수정 (음수를 → 음수는)

StringCaculator.java
- 불필요한 import 제거

ghtjr410 committed:
refactor: Token을 Positive로 리네이밍하여 도메인 의도 명확화

Token.java → Positive.java
- 생성자: int 입력받아 양수 검증
- from(): String 파싱 후 Positive 생성
- toPositive() 제거, value() 접근자 사용

Tokens.java
- List<Token> → List<Positive>
- toPositive() → toIntArray()
- Token::new → Positive::from

StringCalculator.java
- toPositive() 호출부 수정

TokenTest → PositiveTest
- 테스트 전면 수정

ghtjr410 committed:
refactor: Tokens → Positives로 클래스명 및 관련 테스트 클래스명 리네이밍

ghtjr410 committed:
refactor: Positives 책임 강화 및 StringCalculator 단순화

Positives.java
- sum() 메서드 추가
- toIntArray() private으로 변경
- 예외 메시지 "토큰" → "양수" 변경
- 파라미터명 tokens → stringArray 변경

StringCalculator.java
- toPositive(), sum() 중간 메서드 제거
- parse() → split()으로 메서드명 변경
- Positives.from().sum() 직접 호출

PositivesTest.java
- sum() 테스트 추가 (정상/단일/0포함)
- from_음수_포함_문자열_배열 테스트 추가
- 예외 메시지 검증 업데이트

ghtjr410 committed:
refactor: Positive → PositiveOrZero로 클래스명 및 관련 참조 일괄 변경

ghtjr410 committed:
refactor: Positives → PositiveZeros로 클래스명 및 관련 테스트 일괄 변경

ghtjr410 committed:
refactor: Numbers 도메인 클래스 및 관련 테스트 제거

ghtjr410 committed:
refactor: 정적 팩토리 제거 후 생성자 기반으로 로직 단순화

PositiveOrZero.java
- from(String) 정적 팩토리 메서드 제거
- 문자열 입력을 처리하는 생성자 추가로 일관된 객체 생성 방식 적용
- validate 메서드명 변경(validateInvariant)으로 의도 명확화

PositiveZeros.java
- PositiveOrZero 생성 시 새 문자열 생성자 사용하도록 수정

PositiveOrZeroTest.java / PositiveZerosTest.java
- 정적 팩토리 제거에 맞게 테스트를 생성자 기반으로 수정

ghtjr410 committed:
refactor: PositiveZeros → PositiveOrZeros로 리네이밍

ghtjr410 committed:
refactor: 불필요한 static 제거 및 코드 단순화

PositiveOrZero.java
- validateInvariant, isNegative 메서드의 static 제거
- 사용되지 않는 import 문 정리

ghtjr410 committed:
refactor: PositiveOrZeros 생성자 기반으로 구조 단순화

PositiveOrZeros.java
- from() 정적 팩토리 제거 후 생성자(String[], List)로 통합
- 빈 리스트 허용 및 유효성 검증 로직 단순화
- sum 로직 간결화

StringCalculator.java
- PositiveOrZeros.from() 호출을 생성자 방식으로 변경

PositiveOrZerosTest.java
- 새 생성자 방식에 맞게 테스트 수정 및 보강

ghtjr410 committed:
refactor: Pattern 객체를 활용한 커스텀 구분자 접두사 검증

- String.startsWith()를 Pattern.matcher().find()로 변경
- CUSTOM_DELIMITER_PREFIX를 Pattern 객체로 리팩터링
- 정규식 ^// 사용으로 문자열 시작부터 매칭 보장

ghtjr410 committed:
refactor: Matcher 그룹 캡처로 구분자 추출 로직 단순화

StringCalculator.java
- CUSTOM_DELIMITER_PREFIX를 CUSTOM_DELIMITER_PATTERN으로 변경
- 정규식에 그룹 캡처 추가: ^//(.)\\n(.*)
- matcher.group(1), group(2)로 구분자와 숫자 한 번에 추출
- substring 기반 메서드 3개 제거:
  - extractCustomDelimiter()
  - getDelimiterEndPosition()
  - extractNumbersPart() 구현 간소화
- 한 번의 매칭으로 검증과 파싱 동시 처리

ghtjr410 committed:
refactor: 문자열 파싱 로직 분리 및 가독성 개선

StringCalculator.java
- split 메서드를 parse로 변경하고 책임 분리
- 기본/커스텀 구분자 처리 로직을 별도 메서드로 분리
- 정규식 구분자 이스케이프 처리 추가로 안정성 강화

ghtjr410 commented:
피드백 반영했습니다! 

- 정적 팩터리 메서드 제거하고 생성자로 통일했습니다
- Token → PositiveOrZero로 도메인 의미 명확하게 개선했습니다
- Tokens -> PositiveOrZeros로 개선했습니다.
- 테스트 코드도 생성자 기반으로 전면 수정했습니다

추가적으로
- Pattern Matching 활용해서 파싱 로직을 리팩터링했습니다.

말씀하신 피드백들을 어떻게 반영할 수 있을까, 더 개선할 부분은 없을까 
고민하면서 코드 작성하니 시간 가는 줄 모르고 재밌게 했습니다

피드백 부탁드립니다!

javajigi commented on src/main/java/caculator/PositiveOrZero.java:
```diff
@@ -0,0 +1,22 @@
+        }
+    }
+
+    private boolean isNegative(int number) {
+        return number < 0;
+    }
```
```suggestion
    private boolean isNegative() {
        return this.value < 0;
    }
```

PositiveOrZero가 값을 가지고 있는데 굳이 메서드 인자로 전달할 필요가 있을까?

javajigi commented on src/main/java/caculator/PositiveOrZero.java:
```diff
@@ -0,0 +1,22 @@
+    private boolean isNegative(int number) {
+        return number < 0;
+    }
+}
```
```java
    public PositiveOrZero add(int number) {
        return new PositiveOrZero(this.number + number);
    }
```

위와 같이 합을 구하는 로직을 PositiveOrZero가 구현하도록 하면 어떨까?

javajigi replied:
앞의 두 피드백이 객체 지향적인 관점으로 프로그래밍을 하는 사고이니 이번 기회에 한번 고민해 보면 어떨까?

객체를 단순히 값만을 가지는 수동적인 객체가 아니라 능동적이고, 주도적인 객체로 바라보는 시각을 길러보면 좋겠다.

javajigi commented on src/main/java/caculator/PositiveOrZeros.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public PositiveOrZeros(String[] inputs) {
+        this(toValidatedList(inputs));
```
👍

javajigi commented on src/main/java/caculator/PositiveOrZeros.java:
```diff
@@ -0,0 +1,30 @@
+
+    public int sum() {
+        return values.stream()
+                .mapToInt(PositiveOrZero::value)
```
객체의 값을 꺼내지 말고 메시지를 보내 구현한다.

javajigi reviewed (approved):
피드백 반영하느라 넘 수고했어요. 👍 
지금 pr에도 몇 개의 피드백을 남겼는데요.
2단계로 피드백을 계속 주고 받기 보다 자동차 경주 게임 구현하면서 객체 설계 연습을 이어가면 좋을 것 같아요.
피드백 몇개 남겼는데요. 3단계 구현하면서 함께 반영해 보면 좋겠네요.

2단계로 TDD로 구현하기 위해 노력한 것으로 알고 있는데요.
3단계로 TDD로 접근해 보면 좋겠습니다.
